### DISCUSSION TITLE
Plausible on Docker arm64 with outgoing e-mail / smtp - solved!!

### DISCUSSION THREAD

**Discussion by mheland** [plausible/analytics#3134] [Self-Hosted Support]
Docker images for arm64 exists for Plausible, postgres and clickhouse, but not the outgoing mail bytemark/smtp.

Turns out there is a Docker amd64 emulator by Tõnis Tiigi. After pulling tonistiigi/binfmt and assigning `platform: ` in docker-compose.yml it just works. Thank you to  [@otterlord](https://enlear.academy/run-amd64-docker-images-on-an-arm-computer-208929004510) for this.

`docker run --privileged --rm tonistiigi/binfmt --install amd64`

and then...

```
services:
  mail:
    image: bytemark/smtp
    restart: always
    platform: linux/amd64
```

Now I can run Plausible on Hetzner's amd64 Ampere Altra VPS.

Note: Hetzner blocks outgoing smtp ports 25 and 465 until you have been a customer for a month and first invoice is paid - a port 80 proxy or similar to the SMTP host/server is needed initially.

If there is a native arm64 smtp mailer compatible with Plausible pls let me know, I tried a few but couldn't get any to work on the VPS.
/magnus

**Comment by ruslandoga** [comment]
👋 @mheland 

I've been working on removing the need for an SMTP relay but haven't been able to try it out on any hosting providers yet since they all block port 25. I wonder if you could help me try it out? :)

To try it out you'd need to update the image in `docker-compose.yml`

```diff
services:
  plausible:
-   image: plausible/analytics:v2.0.0
+   image: ghcr.io/ruslandoga/analytics:mua-build # it's branched from v2.0.0
```
and configure the new mailing adapter in `plausible-conf.env`
```diff
+ MAILER_ADAPTER=Bamboo.Mua
```

Note that depending on your mail host, you might need to do some extra setup. Like for gmail.com I had to whitelist the host running Plausible in a `TXT` record like `v=spf1 ip4:1.2.3.4 -all` otherwise gmail.com was refusing to accept emails sent from my instance.

More links:

- https://github.com/ruslandoga/mua
- https://github.com/ruslandoga/analytics/pull/126

**Reply by mheland** [reply]
I'd be happy to try it on my dev VPS - I'll give it a try later today or tomorrow.

Fastmail  have anyport-SMTP proxy ;) I'm running SMTP over port 80...

_If you are behind a firewall, some service ports may be blocked. 
Fastmail has a special proxy system whereby you can access IMAP, POP and
SMTP through any port number. All these servers use SSL/TLS._ 

`smtps-proxy.fastmail.com`

_If you use these server names, you can use any port to access these services. Commonly open ports include <code>80</code>, <code>21</code>, <code>25</code> and <code>443</code>._

**Reply by mheland** [reply]
I made it angry... this is from Password Reset on login page

Live at https://pl1.helander.stream/
e-mail redacted...

```
hosting-master-plausible-1            |     Args: []
hosting-master-plausible-1            | 19:41:58.468 [error] Task #PID<0.3873.0> started from #PID<0.3872.0> terminating
hosting-master-plausible-1            | ** (ArgumentError) The recipient `{nil, "user@domain.tld"}` is invalid.
hosting-master-plausible-1            |
hosting-master-plausible-1            | Recipients must be in the format of either a string,
hosting-master-plausible-1            | or a tuple with two elements `{name, address}`
hosting-master-plausible-1            |
hosting-master-plausible-1            |     (mail 0.2.3) lib/mail.ex:436: Mail.validate_recipients/1
hosting-master-plausible-1            |     (mail 0.2.3) lib/mail.ex:278: Mail.put_to/2
hosting-master-plausible-1            |     (bamboo_mua 0.1.0) lib/bamboo_mua.ex:51: Bamboo.Mua.render/1
hosting-master-plausible-1            |     (bamboo_mua 0.1.0) lib/bamboo_mua.ex:11: Bamboo.Mua.deliver/2
hosting-master-plausible-1            |     (bamboo 2.2.0) lib/bamboo/strategies/task_supervisor_strategy.ex:24: anonymous fn/3 in Bamboo.TaskSupervisorStrategy.deliver_later/3
hosting-master-plausible-1            |     (elixir 1.14.3) lib/task/supervised.ex:89: Task.Supervised.invoke_mfa/2
hosting-master-plausible-1            |     (stdlib 4.2) proc_lib.erl:240: :proc_lib.init_p_do_apply/3
hosting-master-plausible-1            | Function: #Function<0.115084270/0 in Bamboo.TaskSupervisorStrategy.deliver_later/3>
hosting-master-plausible-1            |     Args: []

```

**Reply by ruslandoga** [reply]
@mheland Oh, great. The recipient user appears to not have a name. I'll add a check in the Bamboo adapter to encode only the address in that case and rebuild the image.

Fix: https://github.com/ruslandoga/bamboo_mua/pull/2
Update: https://github.com/ruslandoga/analytics/pull/126/commits/a31e56158637e9c8c0d9a5dbe9056433024c4e01
Build: [`ghcr.io/ruslandoga/analytics:mua-build-4f7261f `](https://github.com/ruslandoga/analytics/pkgs/container/analytics)

**Comment by Debcharon** [comment]
I am using ixdotai/smtp:latest in LayerStack's ARM server and it works fine.

```yaml
services:
  mail:
    image: ixdotai/smtp:latest
    restart: always
```
